wagtail/api/shared/filters.py

```diff
@@ -0,0 +1,150 @@
+                    # and tag filtering at the same time is not supported
+                    queryset._filtered_by_tag = True
+                else:
+                    queryset = queryset.filter(**{field_name: value})
+
+        return queryset
+
+
+class OrderingFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        """
+        This applies ordering to the result set
+        Eg: ?order=title
+
+        It also supports reverse ordering
+        Eg: ?order=-title
+
+        And random ordering
+        Eg: ?order=random
+        """
+        if 'order' in request.GET:
+            # Prevent ordering while searching
+            if 'search' in request.GET:
+                raise BadRequestError("ordering with a search query is not supported")
+
+            order_by = request.GET['order']
+
+            # Random ordering
+            if order_by == 'random':
+                # Prevent ordering by random with offset
+                if 'offset' in request.GET:
+                    raise BadRequestError("random ordering with offset is not supported")
+
+                return queryset.order_by('?')
+
+            # Check if reverse ordering is set
+            if order_by.startswith('-'):
+                reverse_order = True
+                order_by = order_by[1:]
+            else:
+                reverse_order = False
+
+            # Add ordering
+            if order_by == 'id' or order_by in view.get_available_fields(queryset.model):
+                queryset = queryset.order_by(order_by)
+            else:
+                # Unknown field
+                raise BadRequestError("cannot order by '%s' (unknown field)" % order_by)
+
+            # Reverse order
+            if reverse_order:
+                queryset = queryset.reverse()
+
+        return queryset
+
+
+class SearchFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        """
+        This performs a full-text search on the result set
+        Eg: ?search=James Joyce
+        """
+        search_enabled = getattr(settings, 'WAGTAILAPI_SEARCH_ENABLED', True)
+
+        if 'search' in request.GET:
+            if not search_enabled:
+                raise BadRequestError("search is disabled")
+
+            # Searching and filtering by tag at the same time is not supported
+            if getattr(queryset, '_filtered_by_tag', False):
+                raise BadRequestError("filtering by tag with a search query is not supported")
+
+            search_query = request.GET['search']
+
+            sb = get_search_backend()
+            queryset = sb.search(search_query, queryset)
+
+        return queryset
+
+
+class ChildOfFilter(BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        if 'child_of' in request.GET:
+            try:
+                parent_page_id = int(request.GET['child_of'])
+                assert parent_page_id >= 0
```

Comment:
> Assertions should *not* be used to test for failure cases that can occur because of **bad user input** or operating system/environment failures, such as a file not being found.

The [python wiki](https://wiki.python.org/moin/UsingAssertionsEffectively) indicates assertions should not be used in this way. I quite like the solution in this instance but I don't know whether in general it's a coding practice we'd like to avoid.